src/main/java/com/mongodb/hibernate/translate/attachment/SQLAstVisitorStateManager.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.translate.attachment;
+
+import static com.mongodb.hibernate.internal.MongoAssertions.assertNotNull;
+import static com.mongodb.hibernate.internal.VisibleForTesting.AccessModifier.PRIVATE;
+import static java.lang.String.format;
+
+import com.mongodb.hibernate.internal.VisibleForTesting;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Manages internal states while translating SQL AST via {@link org.hibernate.sql.ast.SqlAstWalker}.
+ *
+ * <p>{@code SqlAstWalker} is based on {@code visitor} pattern without returning value from the various visitor methods.
+ * The original goal of {@code SqlAstWalker} is for concatenating SQL string internally; but when translating SQL AST
+ * into MQL, it is highly desirable to return value (e.g. {@link com.mongodb.hibernate.translate.mongoast.AstNode}). To
+ * avoid such dilemma, some attachment/detachment mechanism is implemented as follows:
+ *
+ * <ul>
+ *   <li>when some visitor method needs to return some value, it attaches the value to the top {@link Attachment} node
+ *       of a global stack; the node should have been pushed with empty value by some detacher below.
+ *   <li>when some coordinated detacher needs to access the value attached above, it first pushes an {@link Attachment}
+ *       node into global stack with empty value; then invoke some {@link Runnable} instance which is the wrapper of the
+ *       above attacher and then grabs the attached value by popping from the stack.
+ *   <li>both attacher and detacher should provide identical {@link AttachmentKey} to ensure they are on the same pages.
+ * </ul>
+ *
+ * @see org.hibernate.sql.ast.SqlAstWalker
+ * @see AttachmentKey
+ */
+@SuppressWarnings("unchecked")
+public final class SQLAstVisitorStateManager {
+
+    private static final class Attachment<T> {
+        private final AttachmentKey<T> attachmentKey;
+        private @Nullable T attachedValue;
+
+        Attachment(AttachmentKey<T> attachmentKey) {
+            this.attachmentKey = attachmentKey;
+        }
+
+        @Override
+        public String toString() {
+            return "Attachment{" + "attachmentKey=" + attachmentKey + ", attachedValue=" + attachedValue + '}';
+        }
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    final Deque<Attachment<?>> state = new ArrayDeque<>();
```

Comment:
Let's remove the Deque until we prove that it's actually needed.  The design doc doesn't mention it, and it wasn't needed in the POC (which covered a good bit of the query API)

Reply:
yeah, I overthought. The POC implementation is good enough and it covers everything already.